### PR TITLE
Prioritize NavigateTo searches to care most about the current actives docs/project.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This is valuable for people who would like to search, knowing something is in the files they have opened.  Instead of blindly searching everything, we prioritize their opened work, and then search everything else after that.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @jcouv @jinujoseph @sharwell @heejaechang

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs [hunk 1]
```diff
@@ -152,14 +155,31 @@ internal abstract partial class AbstractNavigateToSearchService
...
+            // Prioritize the active documents if we have any.
+            var highPriDocs = priorityDocuments.Where(d => project.ContainsDocument(d.Id))
+                                               .ToImmutableArray();
+
+            var highPriDocsSet = highPriDocs.ToSet();
+            var lowPriDocs = project.Documents.Where(d => !highPriDocs.Contains(d));
+
+            var orderedDocs = highPriDocs.AddRange(lowPriDocs);
+
+#if DEBUG
```
Comment (Neme12) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: Why is this necessary?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: the checks don't have any impact on on the logic of hte fuction.  so i've made them debug-only.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs [hunk 2]
```diff
@@ -152,14 +155,29 @@ internal abstract partial class AbstractNavigateToSearchService
...
             CancellationToken cancellationToken)
         {
             var result = ArrayBuilder<SearchResult>.GetInstance();
-            foreach (var document in project.Documents)
+
+            // Prioritize the active documents if we have any.
+            var highPriDocs = priorityDocuments.Where(d => project.ContainsDocument(d.Id))
+                                               .ToImmutableArray();
+
+            var highPriDocsSet = highPriDocs.ToSet();
```
Comment (jcouv) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: Is `highPriDocsSet ` used?

File: src/Features/Core/Portable/SolutionCrawler/IDocumentTrackingServiceExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+        /// <summary>
+        /// Get a read only collection of all the unique visible documents in the workspace that are
+        /// contained within <paramref name="solution"/>.
+        /// </summary>
+        public static ImmutableArray<Document> GetVisibleDocuments(this IDocumentTrackingService service, Solution solution)
+            => service.GetVisibleDocuments()
+                      .Select(d => solution.GetDocument(d))
+                      .WhereNotNull()
+                      .Distinct()
```
Comment (jcouv) [inline src/Features/Core/Portable/SolutionCrawler/IDocumentTrackingServiceExtensions.cs]: Since we need to de-dupe anyways, would it make sense to use a set for the result? To get the priority documents, it would just be adding one more document to that set. 
Also, would it make sense to use document IDs rather than documents, since that's what seems to be used during search anyways?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SolutionCrawler/IDocumentTrackingServiceExtensions.cs]: So, in teh case where there's an 'active' document **and ** 'visible' documents from the same project, we want the 'active' document to take precedence.  That's why i use a list and not a set.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: whoops.  i intended to use it in computing lowPriDocs to ensure the algorithm was linear and not quadratic.  good catch.

File: src/Features/Core/Portable/SolutionCrawler/IDocumentTrackingServiceExtensions.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SolutionCrawler/IDocumentTrackingServiceExtensions.cs]: > Also, would it make sense to use document IDs rather than documents, since that's what seems to be used during search anyways?

I've flip/flopped about that.  I think i likke it better talking about the real docs, and not IDs...

**Comment by CyrusNajmabadi** [comment]
> I'm not familiar with the testing infrastructure in this area. Is it possible to get some coverage on the new branch (when there is an active document or some visible documents)?

@jcouv I've added some interesting implementations of IDocumentTrackingService in tests to help validate this behavior.

**Comment by jcouv** [comment]
It looks like there's a build error: `D:\j\workspace\windows_relea---a23004dd\src\EditorFeatures\VisualBasicTest\NavigateTo\NavigateToTests.vb(734,31): error BC30516: Overload resolution failed because no accessible 'SetupWorkspace' accepts this number of arguments. [D:\j\workspace\windows_relea---a23004dd\src\EditorFeatures\VisualBasicTest\Microsoft.CodeAnalysis.VisualBasic.EditorFeatures.UnitTests.vbproj]`

**Comment by CyrusNajmabadi** [comment]
@jcouv Fixed.

File: src/EditorFeatures/TestUtilities/NavigateTo/AbstractNavigateToTests.cs [hunk 4]
```diff
@@ -64,7 +64,18 @@ public abstract class AbstractNavigateToTests
 
         internal async Task TestAsync(string content, Func<TestWorkspace, Task> body, bool outOfProcess)
```
Comment (jcouv) [inline src/EditorFeatures/TestUtilities/NavigateTo/AbstractNavigateToTests.cs]: >TestAsync [](start = 28, length = 9)

Very nice. Thanks!

File: src/Workspaces/CoreTestUtilities/ExportProviderCache.cs [hunk 5]
```diff
@@ -163,6 +163,9 @@ bool IsExcludedPart(ComposablePartDefinition part)
             }
         }
 
+        public static ComposableCatalog WithoutPartsOfTypes(this ComposableCatalog catalog, params Type[] types)
```
Comment (jcouv) [inline src/Workspaces/CoreTestUtilities/ExportProviderCache.cs]: >WithoutPartsOfTypes [](start = 40, length = 19)

I couldn't figure out if/where this method is called. Can you confirm it's used?

Comment (CyrusNajmabadi) [reply src/Workspaces/CoreTestUtilities/ExportProviderCache.cs]: It is not.  Sorry, it was a different approach that didn't pan out.

**Comment by jcouv** [comment]
CodeFlow got pretty confused (around iteration 11 or 12, I think), as it showed earlier changes as changes again and it also showed some duplicate files... 
GitHub is displaying the individual commits (11 thru 14) just fine. Not sure what happened.

image

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.cs [hunk 6]
```diff
@@ -18,19 +18,22 @@ internal partial class NavigateToItemProvider : INavigateToItemProvider2
...
+        private readonly IDocumentTrackingService _documentTrackingService;
         private readonly INavigateToItemDisplayFactory _displayFactory;
 
         private CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
 
         public NavigateToItemProvider(
             Workspace workspace,
-            IAsynchronousOperationListener asyncListener)
+            IAsynchronousOperationListener asyncListener,
+            IDocumentTrackingService documentTrackingService)
```
Comment (jcouv) [inline src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.cs]: >documentTrackingService [](start = 37, length = 23)

This parameter is for testing only, right? If so, maybe add a comment.

File: src/EditorFeatures/TestUtilities/NavigateTo/AbstractNavigateToTests.cs [hunk 7]
```diff
@@ -64,7 +64,18 @@ public abstract class AbstractNavigateToTests
...
         {
-            using (var workspace = SetupWorkspace(content))
+            await TestAsync(content, body, outOfProcess, null);
```
Comment (jcouv) [inline src/EditorFeatures/TestUtilities/NavigateTo/AbstractNavigateToTests.cs]: >null [](start = 57, length = 4)

nit: consider naming the `null` argument

File: src/EditorFeatures/TestUtilities/NavigateTo/AbstractNavigateToTests.cs [hunk 8]
```diff
@@ -74,23 +85,30 @@ public abstract class AbstractNavigateToTests
...
         }
 
-        protected TestWorkspace SetupWorkspace(XElement workspaceElement)
+        private protected TestWorkspace SetupWorkspace(
+            XElement workspaceElement, ExportProvider exportProvider, 
+            Func<TestWorkspace, IDocumentTrackingService> createTrackingService)
         {
-            var workspace = TestWorkspace.Create(workspaceElement, exportProvider: TestExportProvider.ExportProviderWithCSharpAndVisualBasic);
-            InitializeWorkspace(workspace);
+            var workspace = TestWorkspace.Create(workspaceElement, exportProvider: exportProvider);
```
Comment (jcouv) [inline src/EditorFeatures/TestUtilities/NavigateTo/AbstractNavigateToTests.cs]: >exportProvider: [](start = 67, length = 15)

nit: consider not naming this argument (if possible)

Comment (heejaechang) [reply src/EditorFeatures/TestUtilities/NavigateTo/AbstractNavigateToTests.cs]: @sharwell is this the way you imagine people inject Mock service to workspace? or some other way? I know you said, in test, we should never use mock object. so not sure whether using lock like this is okay or not.

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs [hunk 9]
```diff
@@ -84,19 +98,80 @@ private class Searcher
                 }
             }
 
-            private async Task SearchAsync(Project project)
+            private async Task SearchProjectsInPriorityOrder(IDocumentTrackingService docTrackingService)
+            {
+                var processedProjects = new HashSet<Project>();
+
+                var activeDocOpt = docTrackingService.GetActiveDocument(_solution);
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs]: > isn't active doc already included in visible docs?

It may be... but it isn't doc'ed to be.  So i wrote with the assumption it might not be.  Also, even if it was, we still want to handle the active doc as *more* important than the visible docs.

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs [hunk 10]
```diff
@@ -84,19 +98,80 @@ private class Searcher
...
+                    var activeProject = activeDocOpt.Project;
+                    processedProjects.Add(activeProject);
+
+                    var visibleDocsFromProject = visibleDocs.Where(d => d.Project == activeProject);
+                    var priorityDocs = ImmutableArray.Create(activeDocOpt).AddRange(visibleDocsFromProject);
+
+                    // Search the active project first.  That way we can deliver results that are
+                    // closer in scope to the user quicker without forcing them to do something like
+                    // NavToInCurrentDoc
+                    await Task.Run(() => SearchAsync(activeProject, priorityDocs), _cancellationToken).ConfigureAwait(false);
```
Comment (heejaechang) [inline src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs]: so, I guess _cancellation is missing in SearchAsync since it is a field? since priorityDocs are not determined by input to Search all, how about making priorityDocs as field as well, and then don't need to pass down on every call?

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs [hunk 11]
```diff
@@ -84,19 +98,80 @@ private class Searcher
                 }
             }
 
-            private async Task SearchAsync(Project project)
+            private async Task SearchProjectsInPriorityOrder(IDocumentTrackingService docTrackingService)
+            {
+                var processedProjects = new HashSet<Project>();
+
+                var activeDocOpt = docTrackingService.GetActiveDocument(_solution);
+                var visibleDocs = docTrackingService.GetVisibleDocuments(_solution)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs]: sorry, i don't understand the question :)  can you calrify?  The reason we don't keep them together is that we want to treat the active-doc as being hte highest priority.

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs]: the priority docs are different *per* project you are searching.  so it can't be a field.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs [hunk 12]
```diff
@@ -152,14 +155,29 @@ internal abstract partial class AbstractNavigateToSearchService
...
             PatternMatcher nameMatcher, PatternMatcher containerMatcherOpt,
             DeclaredSymbolInfoKindSet kinds,
             ArrayBuilder<PatternMatch> nameMatches, ArrayBuilder<PatternMatch> containerMatches,
             CancellationToken cancellationToken)
         {
             var result = ArrayBuilder<SearchResult>.GetInstance();
-            foreach (var document in project.Documents)
+
+            // Prioritize the active documents if we have any.
+            var highPriDocs = priorityDocuments.Where(d => project.ContainsDocument(d.Id))
```
Comment (heejaechang) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: so, this does filtering to find docs that belong to current project. then is there a reason the higher up, caller needs to filter out and create new array for each project?

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs (hunk 10 above)
Comment (heejaechang) [reply src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs]: you changed the list I saw, but looks like that doesn't affect any result. you can give full list everything, and it will still work same.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: sorry, i don't understand hte question.  :)    

>  then is there a reason the higher up, caller needs to filter out and create new array for each project?

What are you referring to?  The caller's job is to say which documents it would prefer be searched first.  This code then uses that to order things properly to respect the caller's wishes.

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (heejaechang) [reply src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs]: I see, so, you want to move the active document to the first one on the list. sure.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs (hunk 12 above)
Comment (heejaechang) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: so you do this per SearchAsync - priorityDocs.ToImmutableArray() - for a project.

I am asking if you create var list = activeDoc,Cocat(restofVisibleDoc) and call all SearchAsync(project, list)

will result or order of documents processed change?

otherwise, we bother to create list again and again?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: This would have different semantics from what i want.  The semantics i want are:

1. process *only* the project with the active document *first*.  Do not do anything else while processing this project.
2. process hte projects for the visible documents in *parallel*.  
3. process anything remaining in parallel.

Placing hte active document at the head of the visible-doc list, and then processing that list in parallel would have a different experience.  Specifically, the most relevant project would now have to compete for CPU time against all the projects that have visible docs.

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs]: Sorry, i'm confused by this:

> you changed the list I saw, but looks like that doesn't affect any result. you can give full list everything, and it will still work same.

THe list is used to determine the order that documents are processed in.  If we pass the full list, then we're effectively saying that no documents should be processed first.  Can you clarify why you think "it will work the same" and won't actually be impacted here?  Thanks!

**Comment by CyrusNajmabadi** [comment]
@jcouv  Ok, i'm happy with this going in as-is.  Thanks!

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs (hunk 12 above)
Comment (heejaechang) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: I dont think any behavior you have currently will change even if you passed in a priority list that contains active document at the head and all other visible document after that. rather than creating new list for each project to filter out documents that belong to one project.

Comment (jcouv) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: I'm waiting for this thread to be resolved, in order to merge the PR. Thanks

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: > I dont think any behavior you have currently will change even if you passed in a priority list 

... why not?  I really don't understand your concern here.  Can you please explain it fully?  If necessary, can you please come to gitter.im so we can discuss this?  I'm really not understanding why you think this would not work.

> that contains active document at the head and all other visible document after that.

As i've explained, we process all visible documents in parallel.  So if the active document is at the front, that means that we'll process it in parallel with the visible documents.  That's *not* what i want.  I want the active document to be processed entirely independently, so it's not sharing any CPU with any other processing.

Comment (jcouv) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: That makes sense to me. 
@heejaechang Are you satisfied too?

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.cs [hunk 13]
```diff
@@ -190,7 +194,7 @@ public ShimNavigateToSearchService(INavigateToSearchService navigateToSearchServ
             public Task<ImmutableArray<INavigateToSearchResult>> SearchDocumentAsync(Document document, string searchPattern, IImmutableSet<string> kinds, CancellationToken cancellationToken)
                 => _navigateToSearchService.SearchDocumentAsync(document, searchPattern, cancellationToken);
 
-            public Task<ImmutableArray<INavigateToSearchResult>> SearchProjectAsync(Project project, string searchPattern, IImmutableSet<string> kinds, CancellationToken cancellationToken)
+            public Task<ImmutableArray<INavigateToSearchResult>> SearchProjectAsync(Project project, ImmutableArray<Document> priorityDocuments, string searchPattern, IImmutableSet<string> kinds, CancellationToken cancellationToken)
                 => _navigateToSearchService.SearchProjectAsync(project, searchPattern, cancellationToken);
```
Comment (heejaechang) [inline src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.cs]: should you use priority document here?

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs (hunk 10 above)
Comment (heejaechang) [reply src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs]: okay, let me try again.

if you change code to this.
var list = activeDocOpt != null ? ImmutableArray.Create(activeDocOpt).AddRange(visibleDocs) : visibleDocs;

and pass the list to every place you have passed in priority documents list. 

and then in ProjectSearch(project, priorityDocuments, ... ) filer out any documents that is not part of the project given.

that way, one doesn't need to create new immutable array per project or do GroupBy and etc. the list can be created at the constructor once so no need to pass it around methods.

like I said, this doesn't change functionality. so if you believe current way is better. sure.

**Comment by heejaechang** [comment]
@jcouv yes I am good.

**Comment by heejaechang** [comment]
retest windows_debug_vs-integration_prtest please

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Searcher.cs]: Ah.  Thank you for the explanation. I see now what you mean.  Yes, i think such an approach could work.

However, my primary issue is simply that i think such an API is fairly confusing.  I would not want to have an API that says: "give priority to these docids in this project".  It makes it really feel like it's a bug that those docids could refer to something *outside* of the project.  It's similar to if an API took a Solution and a Project and hte Project was not from that Solution.

Technically, you could completely document it and say it was "by design", but i still think it's an intuitive thing to do in an API, and i'd prefer the API was crisp and clear here.

Thanks!

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Comment by jcouv** [comment]
Thanks @CyrusNajmabadi. I look forward to using this in a preview.

If I may pique your interest, another thing that is odd with current GoTo behavior is that it finds symbols faster than files. That is especially strange when trying to navigate to an already opened file ;-)

**Comment by sharwell** [comment]
> If I may pique your interest, another thing that is odd with current GoTo behavior is that it finds symbols faster than files. 

That should be fixed in 15.8 Preview 4 and newer. :smile:

**Comment by jcouv** [comment]
@sharwell I hadn't noticed. Cool! I'll keep an eye out.

**Comment by CyrusNajmabadi** [comment]
We're not responsible for file-searching :)  But i'm glad to hear that the core file-name-searcher is now faster :)
